eventauth/eventcontent.go

```diff
@@ -0,0 +1,280 @@
+package eventauth
+
+import (
+	"encoding/json"
+	"strconv"
+	"strings"
+)
+
+// creaateContent is the JSON content of a m.room.create event along with
+// the top level keys needed for auth.
+type createContent struct {
+	// We need the domain of the create event when checking federatebility.
+	senderDomain string
+	// We need the roomID to check that events are in the same room as the create event.
+	roomID string
+	// The "m.federate" flag tells us whether the room can be federated to other servers.
+	Federate *bool `json:"m.federate"`
+	// The creator of the room tells us what the default power levels are.
+	Creator string `json:"creator"`
+}
+
+// load the create event content from the create event in the auth events.
+func (c *createContent) load(authEvents AuthEvents) error {
+	createEvent, err := authEvents.Create()
+	if err != nil {
+		return err
+	}
+	if createEvent == nil {
+		return errorf("missing create event")
+	}
+	if err = json.Unmarshal(createEvent.Content, c); err != nil {
+		return errorf("unparsable create event content: %s", err.Error())
+	}
+	c.roomID = createEvent.RoomID
+	if c.senderDomain, err = domainFromID(createEvent.Sender); err != nil {
+		return err
+	}
+	return nil
+}
+
+// domainAllowed checks whether the domain is allowed in the room by the
+// "m.federate" flag.
+func (c *createContent) domainAllowed(domain string) error {
+	if domain == c.senderDomain {
+		// If the domain matches the domain of the create event
+		return nil
+	}
+	if c.Federate == nil || *c.Federate {
+		// The m.federate field defaults to true.
+		return nil
+	}
+	return errorf("room is unfederatable")
```

Comment:
Yucky, I don't like the style of "filter out successes then what's left is an error" because it is really error-prone when compared to "filter out errors then what's left is a success". For example, what happens in this function if `domain != c.senderDomain` and `c.Federate = true`? The domain will be **allowed** even though the sender domain doesn't match!